src/transaction.ts

```diff
@@ -216,6 +216,160 @@ export const sendTransaction = async function (
   return { transaction, confirmation }
 }
 
+/**
+ * Get all of the unamed resources used by the group in the given ATC
+ *
+ * @param algod The algod client to use for the simulation
+ * @param atc The ATC containing the txn group
+ * @returns The unnamed resources accessed by the group and by each transaction in the group
+ */
+export async function getUnnamedResourcesAccessed(algod: algosdk.Algodv2, atc: algosdk.AtomicTransactionComposer) {
+  const simReq = new algosdk.modelsv2.SimulateRequest({
+    txnGroups: [],
+    allowUnnamedResources: true,
+  })
+
+  const result = await atc.simulate(algod, simReq)
+
+  const groupResponse = result.simulateResponse.txnGroups[0]
+
+  if (groupResponse.failureMessage) {
+    throw Error(`Error during resource packing simulation in transaction ${groupResponse.failedAt}: ${groupResponse.failureMessage}`)
+  }
+
+  return {
+    group: groupResponse.unnamedResourcesAccessed,
+    txns: groupResponse.txnResults.map(
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (t: any) => t.unnamedResourcesAccessed,
+    ) as algosdk.modelsv2.SimulateUnnamedResourcesAccessed[],
+  }
+}
+
+/**
+ * Take an existing ATC and return a new ATC with packed resources
+ *
+ * @param algod The algod client to use for the simulation
+ * @param atc The ATC containing the txn group
+ * @returns A new ATC with the resources packed into the transactions
+ */
+export async function packResources(algod: algosdk.Algodv2, atc: algosdk.AtomicTransactionComposer) {
+  const unnamedResourcesAccessed = await getUnnamedResourcesAccessed(algod, atc)
+  const group = atc.buildGroup()
+
+  unnamedResourcesAccessed.txns.forEach((r, i) => {
+    if (r === undefined) return
+
+    if (r.boxes || r.extraBoxRefs) throw Error('Unexpected boxes at the transaction level')
+    if (r.appLocals) throw Error('Unexpected app local at the transaction level')
+    if (r.assetHoldings) throw Error('Unexpected asset holding at the transaction level')
+
+    // Do accounts first because the account limit is 4
+    r.accounts?.forEach((a) => {
+      group[i].txn.appAccounts?.push(algosdk.decodeAddress(a))
+    })
+
+    r.apps?.forEach((a) => {
+      group[i].txn.appForeignApps?.push(Number(a))
+    })
+
+    r.assets?.forEach((a) => {
+      group[i].txn.appForeignAssets?.push(Number(a))
+    })
+
+    const accounts = group[i].txn.appAccounts?.length || 0
+    if (accounts > 4) throw Error(`Account reference limit of 4 exceeded in transaction ${i}`)
+
+    const assets = group[i].txn.appForeignAssets?.length || 0
+    const apps = group[i].txn.appForeignApps?.length || 0
+    const boxes = group[i].txn.boxes?.length || 0
+
+    if (accounts + assets + apps + boxes > 8) throw Error(`Resource reference limit of 8 exceeded in transaction ${i}`)
```

Comment:
Can we have a const variable for ref limit or is there an const in algosdk that can be used instead of hardcoded 8?

Reply:
I don't think algosdk has those constants but it would really be useful. For now I'll just make constants local to algokit but I think a PR upstream to algosdk with all the consensus parameter tables would be useful.